src/systems/fem_context.C

```diff
@@ -64,78 +66,98 @@ FEMContext::FEMContext (const System &sys)
+      _element_qrule[dim] = hardest_fe_type.default_quadrature_rule
+        (dim, sys.extra_quadrature_order).release();
+      _side_qrule[dim] = hardest_fe_type.default_quadrature_rule
+        (dim-1, sys.extra_quadrature_order).release();
+      if (dim == 3)
+        _edge_qrule = hardest_fe_type.default_quadrature_rule
+          (1, sys.extra_quadrature_order).release();
+
+      // Next, create finite element objects
+      _element_fe_var[dim].resize(nv);
+      _side_fe_var[dim].resize(nv);
+      if (dim == 3)
+        _edge_fe_var.resize(nv);
+
 
-      if ( _element_fe[fe_type] == NULL )
+      for (unsigned int i=0; i != nv; ++i)
         {
-          _element_fe[fe_type] = FEAbstract::build(this->_dim, fe_type).release();
-          _element_fe[fe_type]->attach_quadrature_rule(_element_qrule);
-          _side_fe[fe_type] = FEAbstract::build(this->_dim, fe_type).release();
-          _side_fe[fe_type]->attach_quadrature_rule(_side_qrule);
+          FEType fe_type = sys.variable_type(i);
 
-          if (this->_dim == 3)
+          if ( _element_fe[dim][fe_type] == NULL )
             {
-              _edge_fe[fe_type] = FEAbstract::build(this->_dim, fe_type).release();
-              _edge_fe[fe_type]->attach_quadrature_rule(_edge_qrule);
+              _element_fe[dim][fe_type] = FEAbstract::build(dim, fe_type).release();
+              _element_fe[dim][fe_type]->attach_quadrature_rule(_element_qrule[dim]);
+              _side_fe[dim][fe_type] = FEAbstract::build(dim, fe_type).release();
+              _side_fe[dim][fe_type]->attach_quadrature_rule(_side_qrule[dim]);
+
+              if (this->_dim == 3)
+                {
+                  _edge_fe[fe_type] = FEAbstract::build(dim, fe_type).release();
+                  _edge_fe[fe_type]->attach_quadrature_rule(_edge_qrule);
+                }
             }
-        }
-      _element_fe_var[i] = _element_fe[fe_type];
-      _side_fe_var[i] = _side_fe[fe_type];
-      if (this->_dim == 3)
-        _edge_fe_var[i] = _edge_fe[fe_type];
 
+          _element_fe_var[dim][i] = _element_fe[dim][fe_type];
+          _side_fe_var[dim][i] = _side_fe[dim][fe_type];
+          if ((dim) == 3)
+            _edge_fe_var[i] = _edge_fe[fe_type];
+
+        }
     }
 }
 
-
 FEMContext::~FEMContext()
 {
   // We don't want to store AutoPtrs in STL containers, but we don't
   // want to leak memory either
-  for (std::map<FEType, FEAbstract *>::iterator i = _element_fe.begin();
-       i != _element_fe.end(); ++i)
-    delete i->second;
-  _element_fe.clear();
-
-  for (std::map<FEType, FEAbstract *>::iterator i = _side_fe.begin();
-       i != _side_fe.end(); ++i)
-    delete i->second;
-  _side_fe.clear();
+  for (std::map<unsigned char, std::map<FEType, FEAbstract *> >::iterator d = _element_fe.begin();
+       d != _element_fe.end(); ++d)
+    for (std::map<FEType, FEAbstract *>::iterator i = (d->second).begin();
+         i != (d->second).end(); ++i)
+      delete i->second;
+
```

Comment:
This looks correct, but is scary.  Could you churn a GRINS run through valgrind (after switching map to vector if you take my advice about that) and make sure we're not leaking or mis-deleting anything here?